Watch and reload config file

lclc opened the pull request:
\+ Read config file with Viper (https://github.com/spf13/viper).

With this PR sshmuxd looks for a config file named "sshmuxd" (with any ending you want) in the following places (in this order):

- Applications working dir
- $HOME/.sshmuxd
- /etc/sshmuxd/

When the config file gets changed sshmuxd automatically reloads. If the new config file is not valid JSON it will ignore the changes and logs an error (but keeps running with the old JSON).

My motivation for this PR is so that I can spin-up new instances with Ansible and automatically add them to sshmuxd without needing to restart it.

Viper makes it possible to be config-file-format agnostic, but I only tried with JSON so far. I plan to try with YAML because there is no Ansible role for modifying JSON nicely.

Fixes https://github.com/joushou/sshmuxd/issues/8

I removed passing the filepath as parameter for now. If needed it probably should be easy to add it back. I plan to learn a bit about cobra for CLI interactions in Go next (https://github.com/spf13/cobra), which would I use for sshmuxd too.

Did you consider having hostkey and authkeys also as part of the config file? Might be more convenient to have everything in one file.

Commit:
Watch and reload config file
Read config file with Viper

kennylevinsen commented on main.go:
```diff
@@ -87,20 +59,39 @@ func parseAuthFile(filename string) ([]*sshmux.User, error) {
+	err = viper.UnmarshalKey("hosts", &hosts)
 	if err != nil {
-		panic(err)
+		panic(fmt.Errorf("Error parsing the config file hosts list: %s \n", err))
```
Space before newline

kennylevinsen commented on main.go:
```diff
@@ -87,20 +59,39 @@ func parseAuthFile(filename string) ([]*sshmux.User, error) {
+	viper.AddConfigPath("/etc/sshmuxd/")
+	err := viper.ReadInConfig()
+	if err != nil {
+		panic(fmt.Errorf("Error parsing the config file: %s \n", err))
```
Space before newline

kennylevinsen commented on main.go:
```diff
@@ -87,20 +59,39 @@ func parseAuthFile(filename string) ([]*sshmux.User, error) {
+	viper.SetConfigName("sshmuxd")
+	viper.AddConfigPath(".")
+	viper.AddConfigPath("$HOME/.sshmuxd")
+	viper.AddConfigPath("/etc/sshmuxd/")
```
Could we maintain the option to support user-supplied configuration filename? Something like:

```go
import "flag"
userConfFile := flag.String("configFile", "", "user-supplied configuration file to use")

if userConfFile == "" {
    // Try $HOME and /etc
} else {
    // Use supplied config
}
```

It's kind of convenient for debugging purposes to be able to have more than one configuration file in a directory.

kennylevinsen commented on main.go:
```diff
@@ -87,20 +59,39 @@ func parseAuthFile(filename string) ([]*sshmux.User, error) {
+	})
 
-	hostPrivateKey, err := ioutil.ReadFile(c.HostKey)
+	hostPrivateKey, err := ioutil.ReadFile(viper.GetString("HostKey"))
```
Casing inconsistent with other settings, and with JSON convention.

kennylevinsen commented on main.go:
```diff
@@ -110,13 +101,13 @@ func main() {
 	}
 
-	users, err := parseAuthFile(c.AuthKeys)
+	users, err := parseAuthFile(viper.GetString("AuthKeys"))
```
Casing inconsistent with other settings, and with JSON convention.

kennylevinsen commented on main.go:
```diff
@@ -185,7 +176,7 @@ func main() {
 
 	// Set up listener
-	l, err := net.Listen("tcp", c.Address)
+	l, err := net.Listen("tcp", viper.GetString("Address"))
```
Casing inconsistent with other settings, and with JSON convention.

kennylevinsen reviewed (changes requested):
All in call, awesome, and thanks for your work! Just got a few notes.

lclc commented:
Thanks for the fast feedback and review! Good points, I applied the changes.
I've added the passing for the filename back. It works now with `sshmuxd --config filename.json`

I've also added 2017 in the LICENSE file. Not sure if it is necessary but many major open source projects do it so might not hurt :)

kennylevinsen commented on LICENSE:
```diff
@@ -1,6 +1,6 @@
 The MIT License (MIT)
 
-Copyright (c) 2016 Kenny Levinsen
+Copyright (c) 2016-2017 Kenny Levinsen
```
Could you let it be just "2017" if you want to update it? Not that it legally matters, copyright is entirely implicit, and this text has no effect at all ("all rights reserved" was required many years ago in some obscure countries, but no more). Technically, it's not even correct anymore, as I don't hold the rights to the code you provide without a longer formal legal dance. :)

This is also a really good time to come with a small nugget of wisdom: Focus on the primary task when making a change. It's a little bit silly to have a change blocked or delayed by such a small thing. Keeping PR's (github/bitbucket/...), changesets (gerrit) focused provides huge benefits, both in development efficiency, review efficiency (reviewers don't have to wonder about changes), debugging/testing efficiency (things that break are directly related to the change), version control history and just overall reduce frustrations.

lclc commented:
Sorry, I removed the LICENSE commit.
